File: typescript/src/resources/suppressions.ts
import type { HttpClient } from "../http"
import type {
  AddSuppressionParams,
  Suppression,
  SuppressionListParams,
} from "../types"

export class Suppressions {
  constructor(private http: HttpClient) {}

  /**
   * List suppressions across the company (or a single domain). Suppressed
   * recipients are skipped at send time — every entry here is one address
   * that will not receive mail until removed.
   */
  async list(params?: SuppressionListParams): Promise<Suppression[]> {
    const query: Record<string, unknown> = {}
    if (params?.page !== undefined) query.page = params.page
    if (params?.limit !== undefined) query.limit = params.limit
    if (params?.domainId) query.domainId = params.domainId
    if (params?.reason) query.reason = params.reason
    return this.http.get<Suppression[]>("/suppressions", query)
  }

  /**
   * Manually suppress an address (e.g. honoring an off-platform opt-out).
   * Bounces and complaints are added automatically by the mail server.
   */
  async add(params: AddSuppressionParams): Promise<Suppression> {
    return this.http.post<Suppression>("/suppressions", params)
  }

  /** Remove a suppression — the address will be eligible to receive mail again. */
  async remove(id: string): Promise<void> {
    await this.http.del<{ message: string }>(
      `/suppressions/${encodeURIComponent(id)}`,
    )
  }
}

File: python/sentroy/suppressions.py
from __future__ import annotations

import urllib.parse
from typing import Any, Optional

from sentroy._http import _HttpClient
from sentroy.types import Suppression


class SuppressionsResource:
    """Suppressed recipients — skipped at send time until removed."""

    def __init__(self, http: _HttpClient) -> None:
        self._http = http

    def list(
        self,
        *,
        page: Optional[int] = None,
        limit: Optional[int] = None,
        domain_id: Optional[str] = None,
        reason: Optional[str] = None,
    ) -> list[Suppression]:
        """List suppressions across the company (or a single domain)."""
        query: dict[str, Any] = {}
        if page is not None:
            query["page"] = page
        if limit is not None:
            query["limit"] = limit
        if domain_id:
            query["domainId"] = domain_id
        if reason:
            query["reason"] = reason
        data = self._http.get("/suppressions", query or None)
        return [Suppression.from_dict(s) for s in (data or [])]

    def add(
        self,
        *,
        email: str,
        domain_id: str,
        reason: Optional[str] = None,
    ) -> Suppression:
        """Manually suppress an address (e.g. honoring an off-platform
        opt-out). Bounces and complaints are added automatically by the
        mail server."""
        body: dict[str, Any] = {"email": email, "domainId": domain_id}
        if reason is not None:
            body["reason"] = reason
        data = self._http.post("/suppressions", body)
        return Suppression.from_dict(data)

    def remove(self, id: str) -> None:
        """Remove a suppression — the address becomes eligible to receive
        mail again."""
        self._http.delete(f"/suppressions/{urllib.parse.quote(id, safe='')}")
